### DISCUSSION TITLE
Custom Loaders

### DISCUSSION THREAD

**Discussion by Richienb** [remix-run/remix#3035] [Proposals]
When using a type of file that Remix doesn't support out of the box, the documentation always recommends to run a seperate compiler in parallel with the Remix CLI. For example, to use Sass, I need to [run the Sass compiler to generate css files that Remix can actually import](https://remix.run/docs/en/v1/guides/styling#css-preprocessors).

Then, the [esbuild file loader](https://github.com/remix-run/remix/blob/5b8a0ce0aa0201aa2402fc41405ffbe89605963b/packages/remix-dev/compiler/loaders.ts#L5-L33) appears to be used to get the file into the final build. To simplify the process, an option could be added to the Remix config that allows custom loaders to be specified.

This means, I could just import the sass file where I had previously imported the compiled CSS file and the loader would take care of compiling it. This hands off control of where the intermediate files are stored so I don't need to fiddle with .gitignore and makes it much more convenient to use. Also: [Rollup's Why](https://github.com/rollup/rollup#why).

This could be implemented with a simple config option that internally composes an esbuild plugin. In comparison to just allowing users to [poke around with the esbuild options](https://github.com/remix-run/remix/discussions/2437), a minimal loader API solves a very real problem and only that.

**Comment by sergiodxa** [comment]
This has been asked/suggested before, the Remix team has said they don’t want to expose the compiler or allow plugins right now, the first because that will let them change the internal compiler without causing a breaking change and the second because plugins can slow down your build process and most people will come to this repo asking why is slow instead realizing it’s a plug-in

**Reply by Richienb** [reply]
> that will let them change the internal compiler without causing a breaking change

I'm not asking for fully-featured plugin support. I just want to expand the types of files that can be imported without the maintainers needing to add every one. The API should be incredibly simple so that when the maintainers change the internal API, it would be easy to remap the forward-facing one.

> people will come to this repo asking why is slow instead realizing it’s a plug-in

Build times are already logged so it is not as much of a stretch to include plugin times as well.

**Comment by Richienb** [comment]
I believe the main reason why this has not been added yet is because porting the same feature to different frameworks may turn out difficult. Once Remix comes to Svelte then it would probably be more reasonable since the additional experience would allow for more informed decisions.
